aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -333,6 +333,12 @@ Tensor index_put(const Tensor & self, const torch::List<c10::optional<Tensor>>&
 }
 
 Tensor & _index_put_impl_(Tensor & self, const torch::List<c10::optional<Tensor>>& indices, const Tensor & value, const bool accumulate, const bool unsafe) {
+  if (accumulate && self.device().type() == kCPU) {
```

Comment:
It's worrying that this check and the call to `alertNotDeterministic` occurs even if the global flag is false.

Maybe it'd be more performant to gate on the global flag?

cc @ngimel

Reply:
I don't know whether `device().type()` or checking the flag is faster. Also, there is a deterministic algorithm (use `serial_for_each`) that is being used for small tensors and/or single-threaded runs, so instead of erroring out we should be taking it.

Reply:
I've updated this PR to use the serial implementation when the deterministic flag is enabled, instead of throwing an error.